Guide.shapekey

Mattriks opened the pull request:
<!-- Replace XXX with the issue number that this PR fixes, remove if there is no corresponding issue -->

<!-- Make sure to complete all of these that apply -->
- [x] I've updated the documentation to reflect these changes
- [x] I've added an entry to `NEWS.md`
- [x] I've added and/or updated the unit tests
- [x] I've run the regression tests
- [x] I've `squash`'ed or `fixup`'ed junk commits with git-rebase
- [x] I've built the docs and confirmed these changes don't cause new errors


### `Guide-ageddon`
This is the first step in `Guide-ageddon` (and closes #1133). This PR implements: 
- `Guide.shapekey(title=, labels=, pos=)`
- which generates a colored ShapeKey if `shape` and `color` are mapped to the same variable name. 
- `Theme(key_swatch_color=)`, useful if `shape` and `color` are mapped to different variables.

```julia
srand(123)
theme1 = Theme(point_size=3mm)
coord1 = Coord.cartesian(xmin=0.0, xmax=6.0)
D = DataFrame(x=1:5, y=rand(5), V1=["A","A","B","B","D"], V2 = string.([1,2,2,3,3])  )

pa = plot(x=1:5, y=[0.77, 0.94, 0.67, 0.39, 0.31], shape=["A","A","B","B","D"], theme1, coord1,
    Guide.shapekey(title="Key",labels=["α","β","δ"]),
    Guide.title("Guide.shapekey") )

# if shape==color, and you change levels, there are possible ways to auto-update the levels in Scale.color_discrete
# but for now, you have to do it manually:
pb = plot(D, x=:x, y=:y, shape=:V1, color=:V1,
        Scale.shape_discrete(levels=["D","A","B"]),     
        Scale.color_discrete(levels=["D","A","B"]), 
        theme1, coord1, Guide.title("Shape==Color") )

pc = plot(D, x=:x, y=:y, shape=:V1, color=:V2,  coord1,
        Guide.colorkey(title="Color", pos=[0.75w,-0.25h]),
        Guide.shapekey(title="Shape "),
        Theme(point_size=3mm, key_swatch_color="slategrey"),
        Guide.title("Shape!=Color")
        )
hstack(pa,pb,pc)
```
![shapekey](https://user-images.githubusercontent.com/18226881/40593917-18d74082-626f-11e8-8514-6a6b9f66c985.png)

To do in this PR:  
 - [x] Investigate auto-updating `Scale.color_discrete`, if `shape==color`, and you change `Guide.shapekey(levels=)`  (as in plot b above)

Future PRs (in no particular order):

 - Currently, if you have two keys (e.g. shape and color) positioned outside the plot (left or right) they will stack horizontally not vertically. I leave vertical stacking of left or right keys (and horizontal stacking of top or bottom-positioned keys) to a separate PR.
 - Update `Guide.colorkey`
 - `Guide.sizekey`, `Guide.linekey`
 - Update and replace `Guide.manual_color_key` with `Guide.manual_key`

bjarthur commented:
this will be great to have.  any thoughts on legends for `size_{continuous,discrete}`?

bjarthur commented:
@Mattriks #1116 is ready to merge.  but there will be lots of conflicts with this PR regarding your splitting of `guides.jl`.  i think it would be best to do that in a separate PR.

also, if you think you'll be ready to merge this PR soon, i could wait to merge mine until afterwards.  could work on seaborn-style thumbnails in the meantime.  let me know.

Mattriks commented:
Rebased on master. The tests all pass locally for me. 
Now that I've rebased, I can make the `shapekey` documentation consistent with the new "gallery format".

bjarthur commented:
lgtm except reformatting the docs to the new "gallery format".  let's get this merged before 0.7!

Mattriks commented:
I've updated the docs now, and ran the regression tests.  There are diffs in any testfile that is prefixed "point_shape" or "point_*_shape" i.e. a shape key is added as expected.

bjarthur commented:
thanks!  looking forward to "Future PRs" you list in the OP.

bjarthur commented:
if the length of the shape aesthetic is one would it make sense to not show the shape key?  see http://gadflyjl.org/latest/gallery/shapes.html

Mattriks commented:
Yes, I'll have to fix that, so it is consistent with the behaviour of the `color` aesthetic, e.g. `color=[colorant"orange"]`

Mattriks commented:
Either: 
- let me finish #1175 without the deprecation, or
- show me a way of doing a `Theme` field deprecation which works,
and I'll get back to the above `shape` issue. 🙂 

I also started working on `Guide.linekey`, so I'm thinking about a holistic solution when the aesthetic length is one.

bjarthur commented:
great to hear about `linekey`!

ideally it'd be best if we could fix this regression and tag a new point release before introducing any breaking changes like that in #1175.

tlnagy commented:
I agree with @bjarthur here. I think lets fix this regression first and tag a new point release, merge #1175 and company and push one last feature before we drop support for 0.6 on master.

bjarthur commented:
@Mattriks any chance you'll have time in the near future to get the shape legend to *not* appear when the length of the shape aesthetic is just one?  would be nice to get this fixed before we drop support for julia 0.6.  thanks.
